Demos/DynamicLoader/Loader/main.cpp

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) 2019-2020, Andrew Kaster <andrewdkaster@gmail.com>,
+ * Itamar S. <itamar8910@gmail.com>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+// int foo(int x)
+// {
+//     //
+//     if (x == -1)
+//         foo(0);
+//     return 1;
+// }a
+
+// int libfunc();
+
+#include <dlfcn.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include <AK/LogStream.h>
+#include <AK/String.h>
+#include <LibELF/DynamicLoader.h>
+
+volatile int g_x = 0;
+
+int main(int argc, char** argv, char** envp)
+{
+    (void)argc;
+    (void)argv;
+
+    dbg() << "Loader main";
+    printf("Dynamic loader started\n");
+    char* main_program_path_ptr = getenv("_MAIN_PROGRAM_PATH");
+    ASSERT(main_program_path_ptr);
+    String main_program_path(main_program_path_ptr);
+    char* main_program_fd_str = getenv("_MAIN_PROGRAM_FD");
+    ASSERT(main_program_fd_str);
+    int main_program_fd = atoi(main_program_fd_str);
+    lseek(main_program_fd, 0, SEEK_SET);
```

Comment:
Is this how it works, the loader mmaps the file in itself?

At least on Linux, I believe, both the loader and the executable itself are mmaped in by the kernel.

We do want the loader to be callable directly, as, say, `/path/to/loader /bin/foo bar baz`, so it should keep this manual mmaping functionality, and detect if it's being called by the kernel, as the interpreter, or as the executable.

Reply:
I think the kernel should map both and pass an address to the loader... somehow of where it loaded the main exe (like i said in my comment above :) ). Either ELF Auxiliary vector, or just push it on the stack. Not sure what the kernel will currently try to do when you have a PT_INTERP, been a while since i wrote that part